sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -683,18 +673,22 @@ private static void verifyUnsplittableMethods(ErrorReporter errors, DoFnSignatur
                   i,
                   fnClass.resolveType(params[i]),
                   Arrays.asList(m.getParameterAnnotations()[i])),
-              expectedInputProviderT,
-              expectedOutputReceiverT);
+              inputT,
+              outputT);
 
       methodContext.addParameter(extraParam);
     }
 
     // A splittable DoFn can not have any other extra context parameters.
     if (methodContext.hasRestrictionTrackerParameter()) {
       errors.checkArgument(
-          methodContext.getExtraParameters().size() == 1,
-          "Splittable DoFn must not have any extra arguments, but has: %s",
-          trackerT,
+          Iterables.all(
+              methodContext.getExtraParameters(),
+              Predicates.or(
+                  Predicates.instanceOf(RestrictionTrackerParameter.class),
+                  Predicates.instanceOf(ProcessContextParameter.class))),
+          "Splittable DoFn @ProcessElement must have only ProcessContext and "
```

Comment:
I think this should use more %s and `.class.getSimpleName()` uses to make the message more robust?